events/emitter.go

```diff
@@ -0,0 +1,159 @@
+package events
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/containerd/containerd/api/types/event"
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/gogo/protobuf/proto"
+	protobuf "github.com/gogo/protobuf/types"
+)
+
+const (
+	EventVersion = "v1"
+)
+
+type listener struct {
+	ns string
+	ch chan *event.Envelope
+}
+
+type Emitter struct {
+	listeners map[string]*listener
+}
+
+func NewEmitter() *Emitter {
+	return &Emitter{
+		listeners: make(map[string]*listener),
+	}
+}
+
+func (e *Emitter) Post(ctx context.Context, evt Event) error {
+	for _, l := range e.listeners {
+		topic := getTopic(ctx)
+
+		ns, _ := namespaces.Namespace(ctx)
+		if ns != "" && ns != l.ns {
+			continue
+		}
+
+		eventData, err := convertToAny(evt)
+		if err != nil {
+			log.G(ctx).Error(err)
+			continue
+		}
+
+		l.ch <- &event.Envelope{
+			Timestamp: time.Now(),
+			Topic:     topic,
+			Event:     eventData,
+		}
```

Comment:
It looks like a slow or unresponsive listener will block the other listeners from receiving events. See https://github.com/docker/swarm/issues/2718 for an example of a similar problem in classic Swarm.

Should this use go-events instead so the events are broadcasted to all listeners in a nonblocking fashion?